spec/hyrax/transactions/apply_change_set_spec.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+require 'spec_helper'
+require 'hyrax/transactions'
+require 'dry/container/stub'
+
+RSpec.describe Hyrax::Transactions::ApplyChangeSet do
+  subject(:tx)     { described_class.new }
+  let(:change_set) { Hyrax::ChangeSet.for(resource) }
+  let(:resource)   { build(:hyrax_work) }
+  let(:xmas)       { DateTime.parse('2018-12-25 11:30').iso8601 }
+
+  before { allow(Hyrax::TimeService).to receive(:time_in_utc).and_return(xmas) }
+
+  describe '#call' do
+    it 'is a success' do
+      expect(tx.call(change_set)).to be_success
+    end
+
+    it 'wraps a saved work' do
+      result = tx.call(change_set)
+
+      expect(result.value!).to be_persisted
+    end
+
+    it 'sets modified and uploaded date' do
+      expect(tx.call(change_set).value!)
+        .to have_attributes(date_modified: xmas,
+                            date_uploaded: xmas)
+    end
+
+    context 'with an invalid change_set' do
+      let(:change_set) { change_set_class.new(resource) }
+
+      let(:change_set_class) do
+        Class.new(Hyrax::ChangeSet) do
+          self.fields = [:title]
+
+          validates :title, presence: true
+        end
+      end
+
+      it 'is a failure' do
+        expect(tx.call(change_set)).to be_failure
+      end
+    end
+
+    context 'when save step fails' do
+      before do
+        tx.container.enable_stubs!
+        tx.container.stub('change_set.save', failure_step)
```

Comment:
Now this is some spiffy stubbing behavior. I see that we may wish for a spec_helper for failed containers.

Reply:
pretty nice, huh?

the `Dry::Container` dependency resolution also makes it super easy to inject specialized behavior for any given step at runtime in production.